src/condition/cmdi_detector.cpp

```diff
@@ -0,0 +1,263 @@
+{
+    static const std::string_view whitespaces = " \f\n\r\t\v";
+
+    if (str.empty()) {
+        return str;
+    }
+
+    auto start = str.find_first_not_of(whitespaces);
+    if (start == std::string_view::npos) {
+        return str;
+    }
+
+    auto end = str.find_last_not_of(whitespaces);
+    return str.substr(start, 1 + end - start);
+}
+
+std::size_t object_size(const ddwaf_object &obj) { return static_cast<std::size_t>(obj.nbEntries); }
+
+std::string_view object_at(const ddwaf_object &obj, std::size_t idx)
+{
+    const ddwaf_object &child = obj.array[idx];
+    if (child.type != DDWAF_OBJ_STRING) {
+        return {};
+    }
+    return {child.stringValue, object_size(child)};
+}
+
+std::string_view find_shell_command(std::string_view executable, const ddwaf_object &exec_args)
+{
+    // By initialising the cow_string with the underlying data
+    // contained within the std::string, we ensure a new one won't be
+    // allocated once the string is modified.
+    cow_string executable_lc{executable};
+    if (transformer::lowercase::transform(executable_lc)) {
+        executable = static_cast<std::string_view>(executable_lc);
+    }
+
+    auto shell_it = known_shells.find(basename(executable));
+    if (shell_it != known_shells.end()) {
+        // We've found that the current exec command is attempting to run a
+        // a shell. The shell binary itself might be injected, but also the
+        // shell command. So we need to identify the command
+        std::size_t i = 1;
+        if (!shell_it->second.empty()) {
+            // Most shells allow specifying a command with -c
+            for (; i < object_size(exec_args); ++i) {
+                auto opt = trim_whitespaces(object_at(exec_args, i));
+                if (!opt.empty() && opt[0] == '-' &&
+                    opt.find(shell_it->second) != std::string_view::npos) {
+                    // We've found the -c option, we can now break, if it isn't found
+                    // i will reach the end of the array
+                    break;
+                }
+            }
+        }
+        for (; i < object_size(exec_args); ++i) {
+            auto arg = trim_whitespaces(object_at(exec_args, i));
+            if (arg.empty() || arg.starts_with('-') || arg.starts_with('+')) {
+                continue;
+            }
+
+            // Once the first non-option argument is reached, it must be the
+            // shell command
+            return arg;
+        }
+    }
+    return {};
+}
+
+std::optional<shi_result> cmdi_impl(const ddwaf_object &exec_args,
+    std::vector<shell_token> &resource_tokens, const ddwaf_object &params,
+    const exclusion::object_set_ref &objects_excluded, const object_limits &limits,
+    ddwaf::timer &deadline)
+{
+    const std::string_view executable = trim_whitespaces(object_at(exec_args, 0));
+    if (executable.empty()) {
+        return {};
+    }
+
+    auto shell_command = find_shell_command(executable, exec_args);
```

Comment:
No early bailout in case no command were found?

Reply:
No, because we can still match the executable independently from the shell command.